src/Gren/Kernel/HttpServer.js

```diff
@@ -0,0 +1,90 @@
+/*
+
+import Gren.Kernel.Scheduler exposing (binding, succeed, fail, rawSpawn)
+import HttpServer exposing (ServerError, toRequest)
+import HttpServer.Response as Response exposing (toResponse)
+import Platform exposing (sendToApp, sendToSelf)
+
+*/
+
+const http = require("http");
+
+var _HttpServer_createServer = F2(function (host, port) {
+  return __Scheduler_binding(function (callback) {
+    const server = http.createServer();
+    server.on("error", function (e) {
+      callback(
+        __Scheduler_fail(A2(__HttpServer_ServerError, e.code, e.message)),
+      );
+    });
+    server.listen(port, host, function () {
+      callback(__Scheduler_succeed(server));
+    });
+  });
+});
+
+var _HttpServer_addListener = F3(function (server, router, msg) {
+  server.on("request", function (request, response) {
+    // TODO: support non-http protocols, proxies, and X-Forwarded-For header(s).
+    // Note: the `request` here is a node `http.IncomingMessage`, not a `http.ClientRequest`,
+    // so we can't just look at `request.protocol`, etc.
+    let url = new URL(request.url, `http://${request.headers.host}`);
+    let body = [];
+    request
+      .on("data", function (chunk) {
+        body.push(chunk);
+      })
+      // TODO: Timeouts.
+      // Currently, if the request never ends (because of an error, or...?)
```

Comment:
Could also happen due to spotty networks (mobile, bad coverage) or uploading a large file over a slow network.